universe/readers/event_readers.py

```diff
@@ -0,0 +1,484 @@
+import heapq
+import json
+import logging
+
+from universe import error
+from universe import spaces
+from universe import utils
+from universe.vncdriver import fbs_reader, vnc_proxy_server, vnc_client
+from universe import spaces as vnc_spaces
+
+logger = logging.getLogger(__name__)
+
+class InvalidEventLogfileError(error.Error):
+    pass
+
+
+class FakeTransport(object):
+    def loseConnection(self):
+        pass
+
+    def write(self, data):
+        pass
+
+
+class Factory(object):
+    def __init__(self, error_buffer):
+        self.error_buffer = error_buffer
+        self.deferred = None
+
+
+class MergedEventReader(object):
+    def __init__(self, *timestamped_iterables):
+        """
+        Args:
+            timestamped_iterables: A set of iterables that return dictionaries. Each dictionary must contain a 'timestamp' key that is monotonically increasing
+        """
+
+        self.merged_iterables = heapq.merge(*timestamped_iterables, key=self._extract_timestamp)
+        self.timestamp = None
+
+    @staticmethod
+    def _extract_timestamp(line):
+        if not isinstance(line, dict):
+            raise InvalidEventLogfileError("MultiIterator received a line that's not a dictionary: {}".format(line))
+        try:
+            return float(line['timestamp'])
+        except KeyError:
+            raise InvalidEventLogfileError("MultiIterator received a line without a timestamp: {}".format(line))
+
+    def next(self):
+        return self.__next__()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """
+        Returns the next item from our iterable lists, sorted by timestamp
+        """
+        data = next(self.merged_iterables)
+        timestamp = data['timestamp']
+        if self.timestamp and (timestamp < self.timestamp):
+            raise InvalidEventLogfileError("MultiIterator received a line with an out of order timestamp: {}".format(data))
+
+        self.timestamp = timestamp
+        return data
+
+
+class ActionQueue(object):
+    def __init__(self):
+        self.actions = []
+        self.pixel_format = []
+
+    def key_event(self, key, down):
+        event = spaces.KeyEvent(key, down)
+        self.actions.append(event)
+
+    def pointer_event(self, x, y, buttonmask):
+        event = spaces.PointerEvent(x, y, buttonmask)
+        self.actions.append(event)
+
+    def set_pixel_format(self, server_pixel_format):
+        self.pixel_format.append(server_pixel_format)
+
+    def pop_all(self):
+        output = self.actions, self.pixel_format
+        self.actions = []
+        self.pixel_format = []
+        return output
+
+
+class RewardReader(object):
```

Comment:
There are a lot of common code repeated in these RewardReader, ActionReader, BotactionReader and other reader classes. Consider refactoring it. Consider using inheritance and moving the common code to the base class.